### PR TITLE
Remove Generic Parameter for NodeId

### PR DESCRIPTION
**Pull request by SamuelSchlesinger**
## Description
Remove the generic parameter for NodeID, as it was only used as NodeId everywhere but the tests, but it polluted the entire rest of the node codebase. 

## Risks
I am not sure if this impacts SDKs, smart contracts, or anything else downstream, though my hope is that it is a node-internal change only.

## To Do
* [x] Remove NodeIdT trait
* [x] Remove associated NodeId type from NetworkedReactor trait
* [x] Refactor Sha512 dependence for testing, currently its causing the internals for tls.rs to leak in a way it wasn't before. That said, it had already effectively leaked as the other option I originally pursued was to use openssl directly. There are two options that seem good to me: 1. Extract Sha512 to its own utility crate, use that in the implementation of tls.rs and in testing. 2. Export a testing function from tls.rs or even account.rs for converting PublicKeys into KeyFingerprints (or NodeIds directly, ideally). I was surprised to find this function didn't already exist, though I hope someone comments on this PR that it actually does.
* [x] General self-review and cleanup

### CODE REVIEW

**Comment by SamuelSchlesinger** [comment]
The build is failing due to rustfmt issues, but I run rustfmt on save and I've tried running it on every rust file in the repo in all the ways I know how. Maybe its a rust toolchain issue?

**Comment by SamuelSchlesinger** [comment]
Btw @xcthulhu, in case you see this, the highway (and other) tests pass locally now

**Comment by SamuelSchlesinger** [comment]
> The build is failing due to rustfmt issues, but I run rustfmt on save and I've tried running it on every rust file in the repo in all the ways I know how. Maybe its a rust toolchain issue?

Yep, it was a toolchain issue. I used the exact right format and it fixed it right up. Hopefully build will succeed now.

File: node/src/components/consensus/tests/utils.rs
Comment (marc-casperlabs) [inline node/src/components/consensus/tests/utils.rs]: If we don't need this, I think we can remove it. If we need to reimplement it and do not remember the git history, it's fairly simple to infer an implementation by looking at `ALICE_NODE_ID` :)

Comment (marc-casperlabs) [reply node/src/components/consensus/tests/utils.rs]: If we do keep it, I believe our convention is also to use line comments (which is fairly minor).

File: node/src/tls.rs
Comment (marc-casperlabs) [inline node/src/tls.rs]: ~Can we get away with not making these `pub` somehow? At the very least I believe they should be `pub(crate)` to be in line with the (poorly enforced) style the rest of the code tries to follow.~

Just saw that it is answered in the original description.

**Review by marc-casperlabs — APPROVED** [review]
I am very happy about this, nice job :)

Only some minor nitpicks. I'd be okay merging this without fixing them, but if we have the time, we should look at them.

Small note: You may consider making a feature branch in your own repo, as you'll likely be working on multiple things at the same time.

> Refactor Sha512 dependence for testing, currently its causing the internals for tls.rs to leak in a way it wasn't before. That said, it had already effectively leaked as the other option I originally pursued was to use openssl directly. There are two options that seem good to me: 1. Extract Sha512 to its own utility crate, use that in the implementation of tls.rs and in testing. 2. Export a testing function from tls.rs or even account.rs for converting PublicKeys into KeyFingerprints (or NodeIds directly, ideally). I was surprised to find this function didn't already exist, though I hope someone comments on this PR that it actually does.

These functions might have existed at some point, but were removed because they were not called. The `tls.rs` is very local to the networking layer, we use OpenSSL for TLS only, although the module originally started out with a more library-ish character (and ambition). I'd be fine with re-adding these functions (they may exist in the history of the file).

File: node/src/tls.rs
Comment (SamuelSchlesinger) [reply node/src/tls.rs]: Yep, and to clarify your answer, did you mean to expose the `PublicKey -> KeyFingerprint` function or the `Sha512` newtype?

Comment (SamuelSchlesinger) [reply node/src/tls.rs]: Ah, I think I understand -- look in the history of the file and potentially re-commit the old functions which lived there.

**Comment by SamuelSchlesinger** [comment]
> I am very happy about this, nice job :)

Well thanks! :+1: 
 
> Only some minor nitpicks. I'd be okay merging this without fixing them, but if we have the time, we should look at them.

No rush at all, so we will pick the nits.

> Small note: You may consider making a feature branch in your own repo, as you'll likely be working on multiple things at the same time.

Thanks for the tip!

File: node/src/components/consensus/consensus_protocol.rs
Comment (goral09) [inline node/src/components/consensus/consensus_protocol.rs:234]: We should be able to leave this trait parameterized with `I` and just use `NodeId` in the `era_supervisor.rs` (`ConsensusConstructor`, `EraSupervisor`, etc.). That way `I` is not "polluting" the code and it is abstract where we want it to be - protocol logic.

File: node/src/components/consensus/era_supervisor/era.rs
Comment (goral09) [inline node/src/components/consensus/era_supervisor/era.rs:196]: Ideally, `HighwayProtocol` should still abstract away `NodeId`. It's strictly networking layer type.

File: node/src/components/consensus/consensus_protocol.rs
Comment (SamuelSchlesinger) [reply node/src/components/consensus/consensus_protocol.rs:234]: That is definitely counter to the original assignment I was given, though I'd be happy to understand why you want it to be "abstract" there.

File: node/src/components/consensus/era_supervisor/era.rs
Comment (SamuelSchlesinger) [reply node/src/components/consensus/era_supervisor/era.rs:196]: When you say "its strictly networking layer type", do you mean that, "up to now, it has strictly been a networking layer type", or do you mean that "there are structural reasons why it _must_ be a network only type"?

File: node/src/components/consensus/consensus_protocol.rs
Comment (goral09) [reply node/src/components/consensus/consensus_protocol.rs:234]: Well, truth is the original assignment was pulled out of the hat by Matt and not discussed with the team :) Not saying we should not be doing that entirely, just that the task and its details were not agreed/approved.

File: node/src/components/consensus/era_supervisor/era.rs
Comment (goral09) [reply node/src/components/consensus/era_supervisor/era.rs:196]: I'm saying that it's an abstraction/type/decision of the networking layer and its details are not important for abstraction of consensus protocol.

File: node/src/components/consensus/consensus_protocol.rs
Comment (SamuelSchlesinger) [reply node/src/components/consensus/consensus_protocol.rs:234]: Thank you for the context -- I'll have to think about this for a bit to form an opinion.

Comment (marc-casperlabs) [reply node/src/components/consensus/consensus_protocol.rs:234]: Was it the thing where we would have consensus internally still have a generic `I`, but the rest of the system standardized on `NodeId`?

Comment (goral09) [reply node/src/components/consensus/consensus_protocol.rs:234]: Yes.

Comment (goral09) [reply node/src/components/consensus/consensus_protocol.rs:234]: If you're asking me @marc-casperlabs, then yes - that's what I was thinking about.

**Review by xcthulhu — APPROVED** [review]

**Comment by xcthulhu** [comment]
We don't need this to be generic right now, let's make it generic when we have need for multiple `NodeId`s in consensus.

**Comment by xcthulhu** [comment]
bors merge

**Comment by goral09** [comment]
bors cancel

**Comment by goral09** [comment]
> We don't need this to be generic right now, let's make it generic when we have need for multiple `NodeId`s in consensus.

You’re not making this decision unilaterally Matt. Especially about the code you did not author originally. Let’s wait for other people to chip in.

**Comment by xcthulhu** [comment]
I'll go add @afck to the reviewers, but TBH I remember him saying we could just get rid of it everywhere.  But can't we just make it generic when we need it to be generic?

**Comment by goral09** [comment]
> I'll go add @afck to the reviewers, but TBH I remember him saying we could just get rid of it everywhere. But can't we just make it generic when we need it to be generic?

I wanted to ask Andreas for an opinion as well. I thought I heard him suggesting, earlier today, that we should/could make it generic for Consensus.

Other than that, I think it's a good design decision to not fix consensus code (otherwise pretty generic in terms of types) for a specific node address type.

**Comment by afck** [comment]
At a glance, I'm fine with removing the parameter. I don't know when I'll have a chance to review this PR in detail, but if someone else has, I'm fine with merging it.

**Comment by xcthulhu** [comment]
Let's wait for @fizyk20 to review this too...

**Comment by fizyk20** [comment]
I'll look at the PR in detail tomorrow, but as far as leaving consensus generic goes, I'm leaning towards Mateusz's position. I think consensus shouldn't care about what the ID type is. If this was about making it generic, I'd agree with Matt to make it generic when we need it, but since it's already so, I'd leave it as is (unless this complicates the removal of the generic ID parameter from elsewhere).

File: node/src/effect/requests.rs
Comment (Fraser999) [inline node/src/effect/requests.rs:179]: Looks like this got accidentally deleted?

Same for the `StorageRequest ` below.  I'm not sure of the reason for the `#[repr(u8)]` on it, but I'm guessing that was an accidental deletion too?

File: node/src/tls.rs
Comment (Fraser999) [inline node/src/tls.rs:84]: I see in your PR description that you need these for tests, but I can't see where that's actually happening.

If possible, let's remove these two `pub`s again please.

**Review by Fraser999 — COMMENTED** [review]
Nice work.  I only have a couple of nitpicks, but LGTM otherwise.

File: node/src/tls.rs
Comment (SamuelSchlesinger) [reply node/src/tls.rs:84]: I use it in the lazy `ALICE_NODE_ID`, which is subsequently used in the tests.

File: node/src/effect/requests.rs
Comment (SamuelSchlesinger) [reply node/src/effect/requests.rs:179]: Yep, good catch.

File: node/src/tls.rs
Comment (SamuelSchlesinger) [reply node/src/tls.rs:84]: I could instead expose a function to attempt to produce a `NodeId` from a `PublicKey`, which I think would make sense to have lying around.

**Pull request closed without merging**
